doc,bgpv1: More failure scenario and wording improvement

YutaroHayakawa opened the pull request:
Node failure scenario and working improvement for the existing doc.

```release-note
doc,bgpv1: More failure scenario and wording improvement
```

YutaroHayakawa committed:
docs,bgpv1: A few minor wording improvements

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
docs,bgpv1: Node failure scenario

Add a node failure scenario doc

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Co-authored-by: Harsimran Pabla <128612031+harsimran-pabla@users.noreply.github.com>
Co-authored-by: Ryan Drew <learnitall0@gmail.com>

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+Node Down
+---------
+
+If the node goes down, the BGP session will be lost, and the Cilium datapath
```
```suggestion
If the node goes down, the BGP sessions from this node will be lost, and the Cilium datapath
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+
+If the node goes down, the BGP session will be lost, and the Cilium datapath
+forwarding will be stopped. If the node is shut down gracefully with a proper
+BGP shutdown procedure or TCP connection close, the BGP peer either stops
```
Should we document what is the `proper BGP shutdown procedure` or graceful node shutdown mechanism ?

YutaroHayakawa replied:
I added a new `Node Shutdown` section. Please take a look.

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+forwarding immediately or takes some time to stop forwarding traffic to the
+node depending on the Graceful Restart settings. Since the Cilium datapath is
+stopping, the traffic will be dropped during this period. This is especially
+problematic when you advertise the route to the Service with
```
I think an explanation of why this is problematic is useful. 

BGP peer will continue to forward traffic to this node even if BGP session goes down due to node shutdown till GR restart timer expires.

YutaroHayakawa replied:
Yep, documented it.

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+problematic when you advertise the route to the Service with
+``externalTrafficPolicy=Cluster``.
+
+Mitigation
```
Maybe we can have subsections under this topic. Voluntary node down, Involuntary node down.

YutaroHayakawa replied:
Made a Involuntary Shutdown section. I also added the Voluntary Shutdown section and referred to the Node Shutdown section from it.

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+
+When a node is voluntarily shut down, by gracefully service-out the node, you
+can avoid the packet loss. For example, you can first drain the node to evict
+all workloads. After that, deconfigure the BGP session by removing the label
```
```suggestion
all workloads. After that, deconfigure the BGP sessions by removing CiliumBGPPeeringPolicy node selector label
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+all workloads. After that, deconfigure the BGP session by removing the label
+from the Node object to stop traffic towards the Services with
+``externalTrafficPolicy=Cluster``. Then, wait for a period of time until the
+BGP peer stops forwarding traffic to the node and finally shut down the node.
```
```suggestion
BGP peer removes routes towards the node and finally shut down the node.
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+BGP peer stops forwarding traffic to the node and finally shut down the node.
+
+When a node is involuntarily shut down, there's no direct mitigation. You can
+choose to use or not to use the BGP Graceful Restart feature, depending on the
```
```suggestion
choose to not use the BGP Graceful Restart feature, depending on the
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+
+When a node is involuntarily shut down, there's no direct mitigation. You can
+choose to use or not to use the BGP Graceful Restart feature, depending on the
+trade-off between the failure detection time and stability. Disabling the
```
```suggestion
trade-off between the failure detection time vs stability provided by graceful restart in cases of Cilium pod restarts. Disabling the
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+When a node is involuntarily shut down, there's no direct mitigation. You can
+choose to use or not to use the BGP Graceful Restart feature, depending on the
+trade-off between the failure detection time and stability. Disabling the
+Graceful Restart allows the BGP peer to stop forwarding faster. Even if the
```
```suggestion
Graceful Restart allows the BGP peer to withdraw routes faster. Even if the
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+choose to use or not to use the BGP Graceful Restart feature, depending on the
+trade-off between the failure detection time and stability. Disabling the
+Graceful Restart allows the BGP peer to stop forwarding faster. Even if the
+node is shut down ungracefully, the worst case time to stop forwarding is the
```
```suggestion
node is shut down ungracefully, the worst case time for peer to withdraw routes is the
```

harsimran-pabla commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+Graceful Restart allows the BGP peer to stop forwarding faster. Even if the
+node is shut down ungracefully, the worst case time to stop forwarding is the
+BGP hold time. When the Graceful Restart is enabled, the BGP peer may need hold
+time + restart time to stop forwarding traffic to the node.
```
```suggestion
time + restart time to withdraw routes received from the node.
```

learnitall commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+forwarding immediately or takes some time to stop forwarding traffic to the
+node depending on the Graceful Restart settings. Since the Cilium datapath is
+stopping, the traffic will be dropped during this period. This is especially
+problematic when you advertise the route to the Service with
```
Grammar suggestions:

```suggestion
stopping, traffic will be dropped during this period. This is especially
problematic when you advertise the route to a Service with
```

learnitall commented on Documentation/network/bgp-control-plane-operation.rst:
```diff
@@ -68,6 +68,36 @@ forwarded to a different node than before, will be reset.
+~~~~~~~~~~
+
+When a node is voluntarily shut down, by gracefully service-out the node, you
+can avoid the packet loss. For example, you can first drain the node to evict
```
Grammar suggestions:

```suggestion
When a node is voluntarily shut down, to gracefully service it out, you
can avoid packet loss. For example, you can first drain the node to evict
```

YutaroHayakawa replied:
"by gracefully service it out" should be fine here. Users can avoid packet loss by gracefully shutting down the node.

learnitall commented on Documentation/network/bgp-control-plane-troubleshooting.rst:
```diff
@@ -53,20 +53,19 @@ established because there was a mismatch between the configured
 are likely to appear in the logs, but there are cases where low-level errors,
-such as lack of connectivity to the Peer IP or an eBGP Peer being ahead of the
-1-hop, may not be reflected in the logs. In such cases, using tools like
+such as lack of connectivity to the Peer IP or when eBGP peer is more than 1
```
```suggestion
such as lack of connectivity to the Peer IP or when an eBGP peer is more than 1
```

learnitall reviewed (changes requested):
Great! Thank you for this. I have a couple of small grammar nits, otherwise LGTM.

learnitall reviewed (approved):
Thanks!

YutaroHayakawa committed:
docs,bgpv1: Add Node Shutdown operation guide

Add an operation guide to shut down the node while avoiding packet loss
as much as possible.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

harsimran-pabla reviewed (approved):
Thank you, looks great 🙏
